plugin_runner/sandbox.py

```diff
@@ -23,92 +26,172 @@
 from RestrictedPython.transformer import (
     ALLOWED_FUNC_NAMES,
     FORBIDDEN_FUNC_NAMES,
+    INSPECT_ATTRIBUTES,
     copy_locations,
 )
+from tini import defaultdict
+
+
+def find_submodules(starting_modules: Iterable[str]) -> list[str]:
+    """
+    Given a list of modules, return a list of those modules and their submodules.
+    """
+    submodules = set(starting_modules)
+
+    for module_path in starting_modules:
+        try:
+            module = importlib.import_module(module_path)
+
+            if not hasattr(module, "__path__"):
+                continue
+
+            for _, name, _ in pkgutil.walk_packages(module.__path__, prefix=module.__name__ + "."):
+                submodules.add(name)
+        except Exception as e:
+            print(f"could not import {module_path}: {e}")
+
+    return sorted(submodules)
+
+
+CANVAS_MODULES = (
+    "canvas_sdk.commands",
+    "canvas_sdk.effects",
+    "canvas_sdk.events",
+    "canvas_sdk.handlers",
+    "canvas_sdk.protocols",
+    "canvas_sdk.questionnaires",
+    "canvas_sdk.templates",
+    "canvas_sdk.utils",
+    "canvas_sdk.v1",
+    "canvas_sdk.value_set",
+    "canvas_sdk.views",
+)
+
+CANVAS_SUBMODULES = [
+    found_module
+    for found_module in find_submodules(CANVAS_MODULES)
+    # tests are excluded from the built and distributed module in pyproject.toml
+    if "tests" not in found_module and "test_" not in found_module
+]
+
+ALLOWED_ATTRIBUTES_BY_MODULE = defaultdict(set)
+
+for module_path in CANVAS_SUBMODULES:
+    module = importlib.import_module(module_path)
+
+    allowed_attributes = getattr(module, "__canvas_allowed_attributes__", None)
+
+    if allowed_attributes:
+        ALLOWED_ATTRIBUTES_BY_MODULE[module_path].update(allowed_attributes)
+
+# In use by a current plugin...
+ALLOWED_ATTRIBUTES_BY_MODULE["canvas_sdk.commands"].add("*")
```

Comment:
Should we engage the developer to get them to update their plugin? (can/should be after we release this change, but we should eliminate this deviation once we can)

Reply:
definitely 👍